Make generator build setuptools based

adisbladis opened the pull request:
This makes the generator a regular Python package.

The root level `setup.py` is still there and fills the same role as before, it just does it's installation by making an editable install of the generator, so interactively hacking on the code base works the same as it did before, and if someone was relying on a simple `git pull` for updating that will also still work.

Users are going to have to rebootstrap after this change by calling setup.py again as file paths have changed.
We could add a compat hack in the bat/sh scripts but I felt that it was too ugly for such a small temporary hiccup.

This is a better alternative to https://github.com/Embedded-AMS/EmbeddedProto/pull/55 as distro packages don't need to call into the custom root-level `setup.py`, but can use regular Python packaging approaches.

Version constraints should be loosened up from `generator/setup.py` and the root level `requirements.txt` should be removed.

BartHertog commented:
Hello @adisbladis,
Thank you for your update! I will look into the code in the coming week.
Best regards,
Bart

adisbladis commented:
@BartHertog Ping?

BartHertog commented:
Hello @adisbladis,

This fell lof my radar. It is on it again.

Best regards,

Bart

BartHertog commented:
Hello @adisbladis,

I have been reading your changes, and this is important to get in the repo. As I have never built a Python package before, I have some questions:
1. Is this design meant to be installed in your local virtual environment, or should it also work in a system-wide installment?
2. I was reading the documentation of setuptools. Somewhere I read, they were moving from a setup.py script to a setup.cfg file. Is this possible or does this not work in this case?
3. Under what conditions did you test this? Aka, how can I reproduce your test environment?

Best regards,

Bart

adisbladis commented:
> Is this design meant to be installed in your local virtual environment, or should it also work in a system-wide installment?

Both are supported.

> I was reading the documentation of setuptools. Somewhere I read, they were moving from a setup.py script to a setup.cfg file. Is this possible or does this not work in this case?

Setuptools is moving towards PEP-621 (`pyproject.toml`). If your package is just a normal Python package this is all you need to define.
However `setup.py` is still supported and even required in this case as we customize the build steps.

> Under what conditions did you test this? Aka, how can I reproduce your test environment?

My primary use of this so far has been in my [coffee machine repository](https://github.com/adisbladis/espressos/blob/master/shell.nix#L19-L43) where I use pre-existing [Nix](https://nixos.org/) abstractions to package EmbeddedProto.
Apart from day-to-day usage there I tested the existing venv wrapper in the root for Linux (`protoc-gen-eams`) and ensured they still work as expected.

If you are open to it I could contribute a Github Actions based CI.

adisbladis commented:
> 3\. Aka, how can I reproduce your test environment?

I've added a Github Actions based CI in https://github.com/Embedded-AMS/EmbeddedProto/pull/61 which runs build tests.

adisbladis commented:
@BartHertog Rebased on top of #61.

I've done a few changes since last you check here:
- Moved as much as possible of the setuptools metadata to the declarative `pyproject.toml` file. Only the `protoc` build is now in Python code, everything else is static TOML. 
- Added checks for Python 3.11 to Github Actions.
- Added checks for Protoc 23 to Github Actions.
- Added MacOS compatibility (required a small clang build fix) to the job matrix.

adisbladis commented:
FYI I've also added Windows to the Github Actions matrix in my fork: https://github.com/adisbladis/EmbeddedProto.
I will deal with this in a follow-up.

BartHertog commented:
This looks great.

I've tested this on Linux and on Windows. On Linux the build_tests.sh script works as usual. On Windows I was no not able to generate code. Was this what you were referring to? Otherwise, I will look further into why it was not generating.

Still todo:
1.  I still have to look into the system-wide installation.
2. After this topic is completed I will prepare to release the project via PyPi. Are there any tasks to complete prior to doing this?

adisbladis commented:
> I've tested this on Linux and on Windows. On Linux the build_tests.sh script works as usual. On Windows I was no not able to generate code. Was this what you were referring to? Otherwise, I will look further into why it was not generating.

I was only referring to adding Windows for CI.

> I still have to look into the system-wide installation.

System-wide should work the same as installing into a venv.

> After this topic is completed I will prepare to release the project via PyPi. Are there any tasks to complete prior to doing this?

Afaik, no.

BartHertog commented:
Under Windows on my laptop, I kept running into the problem that embedded_proto_options_pb2.py would not be generated by the setuptools setup.py script. I could not debug this so far and did not get any errors. Today, I tried the branch on a different Windows desktop, everything worked just fine. 

One more thing to test is the system's global installation. Both for Linux and windows.

BartHertog commented:
Something odd happened. All the commits have repeated itself.

adisbladis commented:
They're not repeated. I rebased the PR.

BartHertog commented:
I pushed my merges to your repo. Do you know if this worked?

![image](https://github.com/Embedded-AMS/EmbeddedProto/assets/35932913/7164b581-2421-4494-8980-7bdcfffa0d89)

adisbladis commented:
> That explains it. I thought I pushed my merges to your repo. Do you know if this worked?

Yes, it worked and I rebased _with_ your latest changes.
When rebasing any merge commits on the branch you're rebasing are gone since the commits are replayed as a new nice linear history on top of the new base branch.

BartHertog commented:
I got it working today on my windows system. I removed older versions of Python and installed 3.12.x and surprise it worked. 

Thank you for your help!
